### PR TITLE
[ROCm] Adding a GPU kernel for dropout

### PR DESCRIPTION
**Pull request by ekuznetsov139**
This implements dropout as a single fused kernel. I see performance gains on the order of 5% in BERT-Large compared with the default implementation.

### CODE REVIEW

**Comment by cheshire** [comment]
This is not ROCm-specific, right?
@chsigg could you take a look?

**Comment by cheshire** [comment]
@reedwm

**Comment by ekuznetsov139** [comment]
It is only enabled for ROCm at present, but I expect it would work fine with CUDA as well.

**Comment by reedwm** [comment]
Thank you for the PR! There has been discussion internally on having a fused Dropout op as well. I haven't taken a close look at this PR yet but here is some API feedback.

One tricky aspect is how to pass the mask from the forward op to the backward op. Currently this PR checks if the output is zero in the backward op and if so, assumes it was dropped. But this assumption is not necessarily correct. An output could be zero if the input was zero, in which case the gradient isn't necessarily zero.

A preliminary solution we decided on is to have two ops. The first, called GenerateDropoutMask, would generate a mask, either as a bitarray encoded in an uint8 tensor or as a boolean tensor. The second, Dropout, would take in the mask and do the dropout. Similarly, DropoutGrad would take in the mask. Only GenerateDropoutMask would be stateful.

An alternative is to wrap part of the dropout function with `tf.function(experimental_compile=True)`. This will use XLA to fuse the dropout op automatically. I'm not sure how feasible this is.

In either case, this will likely have a negative performance impact over the current implementation in this PR, which does not materialize the mask on the GPU but instead generates random numbers on the fly. But we need a way to distinguish whether and output was dropped or not when the output is zero.

@alextp, @sanjoy, @ekuznetsov139, or any others, thoughts on one of the two preliminary solutions I mentioned?

**Comment by alextp** [comment]
If you just compile the forward function with XLA we still need to have some mechanism to send the mask to the backwards pass.

I like the idea of splitting dropout into a stateful and a stateless op, so the stateless one is the only one which has gradients. Can we benchmark this alternative?

It might be ok to take a graph with these two ops and no further uses of the mask tensor and rewrite it to use a fused no-gradient-defined op. This way we keep the user code portable and keep the performance, and it's not hard I think to write such a grappler pass.

**Comment by ekuznetsov139** [comment]
Dropout necessarily introduces a degree of randomness, and the impact of failure to transmit the gradient when the input is exactly zero ought to be low, since that would be a rare event (a nonzero value, especially in fp32, is extremely unlikely to become exactly zero through gradient updates, you would expect something like 1 in 2^32 inputs to dropout to be exactly zero.) I did runs of 1M steps of BERT-Large with and without this PR and there was no statistically significant effect on final loss.
After some consideration, my only concern is that training variables may get "stuck" at zero with this logic in some cases. It may be safer to tweak the logic: propagate gradients whenever output is nonzero _or_ input is zero.

**Comment by reedwm** [comment]
Even though it likely won't affect convergence on any models, I think we should still correctly handle the case where the input is exactly 0. A user may pass inputs with 0s when writing unit tests or debugging, and if we treated that case specially, the user could become very confused. Also it's currently unclear what the performance gain would be by handling inputs of 0 incorrectly as compared to having a separate `GenerateDropoutMask` and a `Dropout` op. Depending on the op(s) following Dropout, handing 0s differently may use more memory as it requires keeping the output tensor in memory to pass to DropoutGrad.

If we find we can significantly improve performance even further by handing inputs of 0 differently, we can add an argument to the dropout op to gain the additional performance. But we should first make the normal version of dropout as fast as possible.

@alextp let me know if you disagree with any of this.

> It might be ok to take a graph with these two ops and no further uses of the mask tensor and rewrite it to use a fused no-gradient-defined op. This way we keep the user code portable and keep the performance, and it's not hard I think to write such a grappler pass.

Agreed.

@ekuznetsov139 let me know your thoughts.

**Comment by ekuznetsov139** [comment]
From the performance perspective, both forward and backward kernels are mostly memory-bound. And we need to be concerned with launch overhead. Splitting out GenerateDropoutMask adds an extra kernel launch, and the work running the RNG in it is no longer overlapped with memory accesses, so the forward pass would get substantially slower. On the other hand, the backward pass would be faster, since it'll only need to read 9 bytes per fp32 element (even less if the mask is packed) instead of 12 or 16.

The optimal approach would be to start with the forward kernel I have in the PR, and to make it generate the mask while applying dropout, and then pass that mask instead of the output tensor to ApplyDropoutGrad. 

It may not even be too hard to change the PR. At the Python level, I'm thinking somewhere along these lines
In nn_ops.py:

```
out, mask = gen_nn_ops.dropout(x,rate,noise_shape=noise_shape,seed=seed)
return out
```

in nn_grad.py:

```
def _DropoutGrad(op, grad):
  dx = gen_nn_ops.dropout_grad(grad, op.inputs[1], op.outputs[0], op.outputs[1]) # inputs[1] is rate
```

Would that work?

As to the order of actions, it depends on whether anyone actually intends to write the version with GenerateDropoutMask in the near future. It's one thing to reject a (possibly a bit flawed) solution in the PR in favor of your preferred architecture if that architecture is expected to go in some time soon. It's something else to reject it now and end up with unfused dropout because the alternate solution does not materialize.

**Comment by reedwm** [comment]
The main issue with having the Dropout op return the mask is that it becomes stateful. I'm not sure what the practical disadvantages of this are (@alextp might know) but this is fixed by splitting the op into `GenerateDropoutMask` and `Dropout`. Also @alextp suggested, we can use grappler to turn this into what you suggested, although admittedly this would not work in Eager mode outside tf.functions.

Having `GenerateDropoutMask` return a boolean tensor representing the mask instead of returning the mask from `Dropout` means an extra kernel launch and one more byte per input element needs to be written and read (or bit if we pack the mask). I don't think this would be too bad, although I haven't done any concrete research of measuring the added time this would take compared to the total time of the step in typical models.

In terms of order of actions: I can find someone to work on `GenerateDropoutMask` and implement this, copying parts of your PR (we can add you as a co-author). We first need to decide on the API we want. The reason we don't want to start with a slightly flawed solution first is that we cannot remove an op once we add it. 

> since it'll only need to read 9 bytes per fp32 element (even less if the mask is packed) instead of 12 or 16.

Wouldn't only 5 bytes need to be read for the backwards pass? 4 for the fp32 output gradient, 1 for the mask

**Comment by ekuznetsov139** [comment]
> The main issue with having the Dropout op return the mask is that it becomes stateful.

I'm not sure what it means and why it is an issue. Perhaps someone could elaborate?

>Wouldn't only 5 bytes need to be read for the backwards pass? 4 for the fp32 output gradient, 1 for the mask

Yes, I meant 9 bytes of memory traffic. 5 bytes read, 4 bytes written.

**Comment by sanjoy** [comment]
Maybe this is too naive, but can `GenerateDropoutMask` statefully generate a PRNG *seed* that both the forward and backward passes use?  Then it could run on the host and the PRNG seed could be copied to the device for the fwd and bwd kernels.

**Comment by reedwm** [comment]
> I'm not sure what it means...

Some ops, like [most random ops](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/ops/random_ops.cc;l=28;drc=07eff9985278bf62d3e24f93f9d01b66bade12e5), are set as stateful. The framework uses this information in various ways, like preventing constant folding from removing the op.

> ...and why it is an issue

@alextp can you elaborate? 

Also /CC @sanjoy, would setting `Dropout` as stateful instead of separating it into a stateful `GenerateDropoutMask` op and a non-stateful `Dropout` op have an effect on XLA?

> Would this be alleviated by making the op internal (rename it into _Dropout)?

I think the op would still be considered external if it can be created from the Python side and exported as a GraphDef or SavedModel, and we would still have to permenantly maintain the op. @alextp can you confirm?

**Comment by reedwm** [comment]
> Maybe this is too naive, but can GenerateDropoutMask statefully generate a PRNG seed that both the forward and backward passes use? Then it could run on the host and the PRNG seed could be copied to the device for the fwd and bwd kernels.

Yeah, this also sounds like a good idea and would save memory bandwidth (but not the extra kernel launch). It's a bit confusing from a user's perspective, since the `GenerateDropoutSeed` op would also take in two seeds as all legacy stateful ops do, but this wouldn't exposed to Python users

**Comment by sanjoy** [comment]
> (but not the extra kernel launch)

I was thinking we'd do:

```
u64[2] seed = GenerateDropoutMask() // Runs on host
LaunchKernel(FwdPassKernel, seed[0], seed[1]);
...
LaunchKernel(BwdPassKernel, seed[0], seed[1]);
```

which would still have two kernel launches.

**Comment by sanjoy** [comment]
> Also /CC @sanjoy, would setting `Dropout` as stateful instead of separating it into a stateful `GenerateDropoutMask` op and a non-stateful `Dropout` op have an effect on XLA?

Should work just fine as long as the state is preserved as resource variables.

Please don't put state in `OpKernel` instances. :)

Also CC @cheshire

**Comment by reedwm** [comment]
Unfortunately we would put the state in the OpKernel instances :(. We cannot create a new resource variable without breaking backwards compatibility, since the current `tf.nn.dropout` op does not create variables.

**Comment by sanjoy** [comment]
> We cannot create a new resource variable without breaking backwards compatibility, since the current `tf.nn.dropout` op does not create variables.

Why?  Because they will be part of the checkpoints we generate?  If yes, should we create a new kind of "anonymous" reference variable that has no implications for checkpoints?  Putting state in `OpKernel`s is compiler hostile and I'd avoid it if at all possible.

**Comment by ekuznetsov139** [comment]
> Unfortunately we would put the state in the OpKernel instances :(. We cannot create a new resource variable without breaking backwards compatibility, since the current tf.nn.dropout op does not create variables.

Why not put the state into a tensor, as in my proposal?

**Comment by reedwm** [comment]
> Why? Because they will be part of the checkpoints we generate?

Yes, and if `tf.compat.v1.disable_v2_behavior()` is called, they need to be initialized.

> If yes, should we create a new kind of "anonymous" reference variable that has no implications for checkpoints? Putting state in OpKernels is compiler hostile and I'd avoid it if at all possible.

This could potentially work, but it adds a lot of complexity. It seems better just to go with the original plan of a stateful `GenerateDropoutMask` or `GenerateDropoutSeed` op and a `Dropout` op, which essentially emulates what `tf.nn.dropout` already does but fuses it.

> Why not put the state into a tensor, as in my proposal?

IIUC, your proposal still has state within the OpKernel. In the current PR, you call `random::New64()` with implicitly uses state from the random number generator, so the `Dropout` op should be marked as stateful if we go with your approach.

**Comment by ekuznetsov139** [comment]
> IIUC, your proposal still has state within the OpKernel. In the current PR, you call random::New64() with implicitly uses state from the random number generator, so the Dropout op should be marked as stateful if we go with your approach.

My proposal does not guarantee that saving and reloading the graph would result in the same operation (unless 'seed' is explicitly specified as nonzero). 

If the requirement is to have this kind of reproducibility, then yes, some kind of state needs to be saved somewhere. But, in that case, I don't see how anything changes from splitting Dropout into GenerateDropoutMask+Dropout. Instead of Dropout being stateful, GenerateDropoutMask becomes stateful.

**Comment by gbaned** [comment]
@ekuznetsov139 Can you please resolve conflicts? Thanks!

**Comment by reedwm** [comment]
> My proposal does not guarantee that saving and reloading the graph would result in the same operation (unless 'seed' is explicitly specified as nonzero).

I don't think it needs to meet that requirement. However, if the kernel uses state, it must be marked as stateful. One reason for this is to prevent certain optimizations. For example, you wouldn't want grappler to use constant propagation to remove the Dropout layer, since then the output of the layer would be the same in every case.

> @ekuznetsov139 Can you please resolve conflicts? Thanks!

No need to do so until we resolve what the next steps are.

@ekuznetsov139 would you want to change this PR to implement and use the `GenerateDropoutMask` or `GenerateDropoutSeed` op? If not, I can find someone internally to work on this. In either case, we first need to decide which of the two ops to go with.

**Comment by ekuznetsov139** [comment]
How about I just add the IsStateful tag to the Dropout op as implemented?

**Comment by reedwm** [comment]
The issue with adding IsStateful to the Dropout op is that we have to maintain this Dropout op indefinitely. We'll then later create a `GenerateDropoutMask` and `DropoutV2` op but we'll still have to maintain the original op.

**Comment by ekuznetsov139** [comment]
Why would you want to create GenerateDropoutMask later? What is there to be gained from having this (slower) two-step process?

**Comment by gbaned** [comment]
@reedwm Can you please assist on above comments from @ekuznetsov139. Thanks!

**Comment by alextp** [comment]
@ekuznetsov139 we discussed why a GenerateDropoutMask is useful in a few messages in this thread. For example it allows you to use the same dropout mask across all cells of an RNN, whether it's unrolled statically (with a python for loop without autograph) or dynamically (with tf.while_loop). It also allows for more reproducible experiments, which is important for some of the researchers in our audience.

We should also try to find a way of making all state used by TF serializable somewhere. Ideally this can be done using the new tf.random.Generator API (which can be an argument to the GenerateDropoutMask op).

**Comment by ekuznetsov139** [comment]
The op as implemented takes a seed as one of the inputs. It only falls back on the internal RNG if the seed is zero. If and when these use cases are covered, the op will not need to be changed - you just need to implement GenerateDropoutSeed, add a call to it into python code, and pass the result to the dropout op.

**Comment by gbaned** [comment]
@ekuznetsov139 Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@ekuznetsov139  Any update on this PR? Please. Thanks!

**Comment by ekuznetsov139** [comment]
@gbaned at this point it is totally up to @reedwm and @alextp. Sounds like they don't want this architecture and they either want to reject it or they want me to implement GenerateDropoutSeed, which I can't do unless I have some directions (the intended API interface in particular).

**Comment by gbaned** [comment]
@ekuznetsov139 Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@reedwm, @alextp  Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@reedwm, @alextp Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@reedwm, @alextp  Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@reedwm, @alextp Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@reedwm, @alextp Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@reedwm, @alextp Any update on this PR? Please. Thanks!
